File: grind-master-dsa/1.two-sum/Q1.js
/*
* Given an array of integers nums and an integer target, 
* return indices of the two numbers such that 
* they add up to target.
*/

const nums = [6, 3, 1, 5, 2];
const target = 7;

console.log(twoSumMethod1(nums, target));
console.log(twoSumMethod2(nums, target));

function twoSumMethod1(nums, target) {
    const indices = [-1, -1];
    const size = nums.length;

    for (let i = 0; i < size; i++) {
        for (let j = i + 1; j < size; j++) {
            const sum = nums[i] + nums[j];
            if (sum === target) {
                indices[0] = i;
                indices[1] = j;
                return indices;
            }
        }
    }

    return indices;
}

function twoSumMethod2(nums, target) {
    const indices = [-1, -1];
    const lookup = new Map();
    const size = nums.length;

    for (let i = 0; i < size; i++) {
        const difference = target - nums[i];
        if (lookup.has(difference)) {
            indices[0] = lookup.get(difference);
            indices[1] = i;
            return indices;
        }
        lookup.set(nums[i], i);
    }

    return indices;
}

File: grind-master-dsa/1.two-sum/Q1.kt
/*
* Given an array of integers nums and an integer target, 
* return indices of the two numbers such that 
* they add up to target.
*/

fun main() {
    var nums = arrayOf<Int>(6, 3, 1, 5, 2);
    var target = 7;

    println(twoSumMethod1(nums, target).joinToString());
    println(twoSumMethod2(nums, target).joinToString());
}

fun twoSumMethod1(nums: Array<Int>, target: Int): Array<Int> {
    var indices = arrayOf<Int>(-1, -1);

    for ((i, v) in nums.withIndex()) {
        for ((j, m) in nums.slice(i + 1..nums.size - 1).withIndex()) {
            var sum = v + m;
            if (sum == target) {
                indices[0] = i;
                indices[1] = j + i + 1;
                return indices;
            }
        }
    }

    return indices;
}

fun twoSumMethod2(nums: Array<Int>, target: Int): Array<Int> {
    var indices = arrayOf<Int>(-1, -1);
    val lookup: HashMap<Int, Int> = HashMap();

    for ((i, v) in nums.withIndex()) {
        var neededNumber = target - v;

        if (lookup.containsKey(neededNumber)) {
            indices[0] = lookup.getOrDefault(neededNumber, -1);
            indices[1] = i;
            return indices;
        }

        lookup[v] = i;
    }

    return indices;
}